doc/developer/design/19042023_stash_migrations.md

```diff
@@ -0,0 +1,352 @@
+    * I wrote a PoC procedural macro that we could use on our Stash objects that would 
+    automatically snapshot them. You could annotate fields in a struct with `#[stash(introduced = 5)]`
+    and that would generate all versions of the struct >= 5 with that field, [branch](https://github.com/parker-timmerman/materialize/commit/243d76fb123625121949155ee5168b50f96cb1e4).
+    * ❌ The reason I don't like this approach is it essentially requires us to maintain a bespoke
+    code generator, built with a seldomly used feature of Rust, procedural macros.
+* Using schema-less `serde_json::Value` in the migration code path, to avoid snapshotting 
+altogether.
+    * Instead of snapshotting our types, we could build a migration path that uses "untyped"
+    `serde_json::Value` structs, where we'd check and modify the fields at runtime.
+    * ❌ We define a [`trait Data`](https://github.com/MaterializeInc/materialize/blob/23d7cad1211b4d1d1de3a26652937f9f2c4df61c/src/stash/src/lib.rs#L101-L103)
+    that all of our keys and values must implement. A requirement of `Data` is that the type must
+    implement [`Ord`](https://doc.rust-lang.org/std/cmp/trait.Ord.html), which seems to be required
+    for consolidating rows. `serde_json::Value` is not order-able, and there isn't a good way to
+    define how a generic JSON value should be ordered.
+* Maintaining `objects_vX.rs` with Rust types instead of an `objects_vX.proto`.
+    * Instead of defining our Stash types in protobuf we could keep them defined in Rust and follow
+    the same approach for snapshotting the `objects_vX.rs` files these types would be in. 
+    * ❌ An issue with this approach is "snapshot hygiene". Nothing would prevent us from adding 
+    dependencies to these snapshots, and to be correct we'd also need to also snapshot the
+    dependency. But we don't have a tool to assert dependencies are snapshotted, so a dependency
+    could change which could silently introduce a change to the serialized format, causing bugs.
+
+### Why this approach?
+
+I believe this is the right design because of the following reasons:
+
+* It allows us to write easy to reason about migration functions that convert from `TypeA` to `TypeB`.
+* Introduces automated checks to make sure a migration exists whenever Stash objects change.
+* Old migrations do not need to change when new fields are introduced, since they can operate on
+old versions of the types.
+* Introduces the possibility of simulation testing for upgrades, since we maintain the history of
+the schema of the Stash, through snapshotting.
+
+Overall I believe it achieves the goal of "fearless migrations", for the lowest possible
+engineering cost.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* How important do we believe human readability of the Stash is?
+* Does the Stash need to be JSON for billing purposes? This [issue](https://github.com/MaterializeInc/materialize/issues/14264)
```

Comment:
It does not! I feel very strongly that nothing but the one `environmentd` in charge should be looking at a given stash, with an exception for emergency debugging tools (`stash-debug`). If you want to know what's in a stash, ask `environmentd`; don't look in the stash directly. We changed the design of billing to adhere to this principle.

Reply:
Great! I totally agree with the idea that `environmentd` should be the only application code that interacts with the Stash, let me call this out in the doc